src/net.cpp

```diff
@@ -1801,9 +1801,8 @@ void CConnman::ThreadDNSAddressSeed()
             unsigned int nMaxIPs = 256; // Limits number of IPs learned from a DNS seed
             if (LookupHost(host, vIPs, nMaxIPs, true)) {
                 for (const CNetAddr& ip : vIPs) {
-                    int nOneDay = 24*3600;
                     CAddress addr = CAddress(CService(ip, Params().GetDefaultPort()), requiredServiceBits);
-                    addr.nTime = GetTime() - 3*nOneDay - rng.randrange(4*nOneDay); // use a random age between 3 and 7 days old
+                    addr.nTime = Now<TimeSeconds>() - 3 * 24h - rng.rand_seconds(4 * 24h); // use a random age between 3 and 7 days old
```

Comment:
Maybe `rand_uniform_delay(Now<TimeSeconds>() - 7*24h, /*range=*/4*24h)` would be better?

Reply:
> You can pick up the precision from the first argument that way

I don't think this is possible. If you want a uniform delay in `[2h, 3h)`, typing `rand_uniform_delay(2h, 1h)` will produce the *wrong* result.

Reply:
This will still break if you call `tp_2 = rand_uniform_delay(tp_1, 1h)` and `tp_1` is of type `TimeHours` in this context, and `tp_2` of type `TimeSeconds`. If you want to make this templated, I think the `std::common_type` approach that `GetRandDur` uses is the only way. Though, as this forces you to specify the precision, you might as well put the precision right into the function name.

Moreover, I am not sure if we should force time points here. After all this is just a duration and it is up to the caller to apply it to a time point or further use it as a duration.

So I think all I can do here is rename `rand_seconds` to `rand_uniform_seconds`.

Reply:
Using `TimeHours` would already be a mistake: if you're measuring time only with a precision of hours, you've already got +/- 30min error to whatever real time you're trying to target.

We should be using full precision time points (`MockClock::time_point`) everywhere we can; only reason not to do that for `addr.nTime` is that we only have seconds since the epoch in the serialization format, and using higher precision data when we throw it through the hash on the first run and lower precision data on later runs would be confusing.

Reply:
Hmmm, one way to implement "I want to run this exactly once per hour, but at a uniform random time within that hour" is to have `tp_2` to be the execution time in seconds and `tp_1` be `tp_2` cast to hours `+ 1h`.

The code with `rand_uniform_delay` would look like:

```
tp_2 = rand_uniform_delay(std::chrono::time_point_cast<TimeSeconds>(std::chrono::time_point_cast<TimeHours>(tp_2 + 1h)), 1h)
```

or 

```
tp_2 = rand_uniform_delay<TimeSeconds>(std::chrono::time_point_cast<TimeHours>(tp_2 + 1h), 1h) // not sure if this compiles
```

Whereas with `rand_uniform_seconds` it would look like:

```
tp_2 = rand_uniform_seconds(std::chrono::time_point_cast<TimeHours>(tp_2 + 1h), 1h)
```

or (if durations are used)

```
tp_2 = std::chrono::time_point_cast<TimeHours>(tp_2 + 1h) + rand_uniform_seconds(0h, 1h)
```

I am not sure how often quantization is used with random delays, but I wouldn't call it a "mistake" to intentionally quantize time.

Absent a use case for quantization, `rand_uniform_seconds` still is a bit less code/less `<>`. Moreover it can be used for non-time-point "plain" durations. So I have a slight preference for it. If there was a use case for quantization, it would also be safer.

Reply:
> Hmmm, one way to implement "I want to run this exactly once per hour, but at a uniform random time within that hour" is to have `tp_2` to be the execution time in seconds and `tp_1` be `tp_2` cast to hours `+ 1h`.

Is there anywhere that we do that? Everything I've come across is a variation on `GetTime() + random(range) + offset`.

I'd probably prefer to write that as:

```c++
when = rand_uniform_delay(round_up(now(), 1h), 1h);
   
template<typename TP>
TP round_up(TP tp, typename TP::duration dur)
{
    return tp + dur - (tp.time_since_epoch() % dur);
}
```

since then you don't have to declare a new type if you want things every 90 minutes instead of every hour. Don't think you can get a compile time check that `dur > 0` though (at least with C++17)

> ```
> tp_2 = rand_uniform_seconds(std::chrono::time_point_cast<TimeHours>(tp_2 + 1h), 1h)
> ```

I guess I don't really think you should ever want/need to use a `time_point_cast` outside of util/time.h/cpp.

> I am not sure how often quantization is used with random delays, but I wouldn't call it a "mistake" to intentionally quantize time.

No, the mistake's doing it via the `time_point` object. The whole benefit of durations and time_points is the compiler can automatically deal with you specifying things in minutes or hours without having to do manual conversions all the time.

> Absent a use case for quantization, `rand_uniform_seconds` still is a bit less code/less `<>`.

Maybe I'm misunderstanding what `rand_uniform_seconds` is, but:

```c++
auto rand_seconds(std::chrono::seconds max) noexcept { return decltype(max){randrange(max.count())}; }
auto rand_millis(std::chrono::milliseconds max) noexcept { return decltype(max){randrange(max.count())}; }
auto rand_micros(std::chrono::microseconds max) noexcept { return decltype(max){randrange(max.count())}; }
```

doesn't seem like notably less code than:

```c++
template<typename TP>
auto rand_uniform_delay(TP tp, typename TP::duration range) noexcept
{ 
    return tp + (typename TP::duration){randrange(range.count())};
}
```

> Moreover it can be used for non-time-point "plain" durations. So I have a slight preference for it. If there was a use case for quantization, it would also be safer.

I think all the current uses of randrange against chrono inputs are getting added to `GetTime()` of one sort or another, so I don't think there'd be a use case for that once the conversion to time_points is done.

Reply:
`scheduleFromNow` is already an inline function that takes a duration and adds it to a time point in order to call `schedule`.

Reply:
Ah good point. I'll take a stab at this soon(TM)

Reply:
Thanks, fixed and resolved conversation.